### PR TITLE
build: assign region at runtime in e2e tests

### PR DESCRIPTION
**Pull request by sobolk**
#### Description of changes

This PR's goal is to remove as much of `CLI_REGION` assignment from generated e2e workflow as possible to free up space (cb yml files have hard size limit).
Therefore:
1. Split e2e test script does not assign region unless test is force-pinned to some region.
2. Adds a script that assigns region at runtime using job identifier.

#### Description of how you validated changes

E2E test run

From handful of jobs:

image

image

image

image

image

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: scripts/select-region-for-e2e-test.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * In order to reshuffle regions CLI_REGION_SALT should be modified in workflow yml definition.
+ *
+ * If region is already assigned, i.e. CLI_REGION environment variable is set this script is pass-through.
+ */
+
+const toNumber = (buf: Buffer) => {
+  return buf.readUInt16BE(0) * 0xffffffff + buf.readUInt32BE(2);
+};
+
+// Algorithm from https://github.com/watson/hash-index/blob/fd5d0606926a821166e428b6e266f03b4dc5d817/index.js#L19
```
Comment (sobolk) [inline scripts/select-region-for-e2e-test.ts:17]: Decided to copy as algorithm is small and library has not been released in years (it doesn't have deps and is simple so no wonder there).

**Review by edwardfoyle — APPROVED** [review]

File: scripts/select-region-for-e2e-test.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+let selectedRegion = process.env.CLI_REGION;
+if (!selectedRegion) {
+  const testSuite = process.env.TEST_SUITE;
+  if (!testSuite) {
+    throw Error('TEST_SUITE environment variable must be set');
+  }
+  // See https://en.wikipedia.org/wiki/Salt_(cryptography).
+  // Salt should be changed if we want re-shuffle regions but keep generation deterministic.
+  const salt = 'foo';
+  const regionIndex = computeHashIndex(`${testSuite}${salt}`, AWS_REGIONS_TO_RUN_TESTS.length);
```
Comment (awsluja) [inline scripts/select-region-for-e2e-test.ts:32]: what kind of distribution do we get, is it an even distribution across the regions?

Comment (awsluja) [reply scripts/select-region-for-e2e-test.ts:32]: what if we take the job name, and its index in the 'wait_for_ids.json' file, to create an even distribution?

ie,
indexOfJobInJobsList = wait_for_ids.find(..) + randomsalt;
CLI_REGION = REGIONS[indexOfJobInJobsList % REGIONS.length]


similarly, we can shuffle regions by shifting the random salt

code is probably off, but general idea

Comment (sobolk) [reply scripts/select-region-for-e2e-test.ts:32]: The hashing function as is seems to assign regions pretty evenly.

image

image


The `wait_for_ids.json` is interesting option though. I'll give it a try.

**Review by edwardfoyle — DISMISSED** [review]

**Review by awsluja — DISMISSED** [review]

**Review by awsluja — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

**Comment by sobolk** [comment]
finally 
image

**Pull request merged**